### PR TITLE
[KYUUBI #5534][AUTHZ] Support optimize table command for Delta Lake

### LINKED ISSUE #5534: [Subtask] Support optimize table command for Delta Lake
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues) and found no similar issues.

### Describe the subtask

Support optimize table command for Delta Lake in Authz plugin.

### Parent issue

#5477 

### Are you willing to submit PR?

- [X] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to improve.
- [ ] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by zml1206**
### _Why are the changes needed?_
To close #5534 .
Support optimize into table command for Delta Lake. This feature is available in Delta Lake 1.2.0 and above.
https://docs.delta.io/latest/optimizations-oss.html#optimize-performance-with-file-management

Delta Lake 3.0/Spark 3.5
```
== Analyzed Logical Plan ==
path: string, metrics: struct<numFilesAdded:bigint,numFilesRemoved:bigint,filesAdded:struct<min:bigint,max:bigint,avg:double,totalFiles:bigint,totalSize:bigint>,filesRemoved:struct<min:bigint,max:bigint,avg:double,totalFiles:bigint,totalSize:bigint>,partitionsOptimized:bigint,zOrderStats:struct<strategyName:string,inputCubeFiles:struct<num:bigint,size:bigint>,inputOtherFiles:struct<num:bigint,size:bigint>,inputNumCubes:bigint,mergedFiles:struct<num:bigint,size:bigint>,numOutputCubes:bigint,mergedNumCubes:bigint>,numBatches:bigint,totalConsideredFiles:bigint,totalFilesSkipped:bigint,preserveInsertionOrder:boolean,numFilesSkippedToReduceWriteAmplification:bigint,numBytesSkippedToReduceWriteAmplification:bigint,startTimeMs:bigint,endTimeMs:bigint,totalClusterParallelism:bigint,totalScheduledTasks:bigint,autoCompactParallelismStats:struct<maxClusterActiveParallelism:bigint,minClusterActiveParallelism:bigint,maxSessionActiveParallelism:bigint,minSessionActiveParallelism:bigint>,deletionVectorStats:struct<numDeletionVectorsRemoved:bigint,numDeletionVectorRowsRemoved:bigint>,numTableColumns:bigint,numTableColumnsWithStats:bigint>
OptimizeTableCommand DeltaOptimizeContext(false,None,None)
+- ResolvedTable org.apache.spark.sql.delta.catalog.DeltaCatalog@6ca257de, delta_ns.table1_delta, DeltaTableV2(org.apache.spark.sql.SparkSession@ae1e520,file:/var/folders/gc/c__qhntd7s502txfp0ltxh880000gn/T/spark-warehouse-fb7cc4a2-7255-4da2-9e5e-a4e2c98c10e0/delta_ns.db/table1_delta,Some(CatalogTable(
Catalog: spark_catalog
Database: delta_ns
Table: table1_delta
Owner: default_table_owner
Created Time: Tue Nov 07 19:52:20 CST 2023
Last Access: UNKNOWN
Created By: Spark 3.5.0
Type: MANAGED
Provider: delta
Location: file:/var/folders/gc/c__qhntd7s502txfp0ltxh880000gn/T/spark-warehouse-fb7cc4a2-7255-4da2-9e5e-a4e2c98c10e0/delta_ns.db/table1_delta
Serde Library: org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe
InputFormat: org.apache.hadoop.mapred.SequenceFileInputFormat
OutputFormat: org.apache.hadoop.hive.ql.io.HiveSequenceFileOutputFormat
Partition Provider: Catalog)),Some(delta_ns.table1_delta),None,Map()), [id#830, name#831, gender#832, birthDate#833]
```
before 
```
== Analyzed Logical Plan ==
path: string, metrics: struct<numFilesAdded:bigint,numFilesRemoved:bigint,filesAdded:struct<min:bigint,max:bigint,avg:double,totalFiles:bigint,totalSize:bigint>,filesRemoved:struct<min:bigint,max:bigint,avg:double,totalFiles:bigint,totalSize:bigint>,partitionsOptimized:bigint,zOrderStats:struct<strategyName:string,inputCubeFiles:struct<num:bigint,size:bigint>,inputOtherFiles:struct<num:bigint,size:bigint>,inputNumCubes:bigint,mergedFiles:struct<num:bigint,size:bigint>,numOutputCubes:bigint,mergedNumCubes:bigint>,numBatches:bigint,totalConsideredFiles:bigint,totalFilesSkipped:bigint,preserveInsertionOrder:boolean,numFilesSkippedToReduceWriteAmplification:bigint,numBytesSkippedToReduceWriteAmplification:bigint,startTimeMs:bigint,endTimeMs:bigint,totalClusterParallelism:bigint,totalScheduledTasks:bigint,autoCompactParallelismStats:struct<maxClusterActiveParallelism:bigint,minClusterActiveParallelism:bigint,maxSessionActiveParallelism:bigint,minSessionActiveParallelism:bigint>,deletionVectorStats:struct<numDeletionVectorsRemoved:bigint,numDeletionVectorRowsRemoved:bigint>,numTableColumns:bigint,numTableColumnsWithStats:bigint>
OptimizeTableCommand `delta_ns`.`table1_delta`, DeltaOptimizeContext(false,None,None)
```

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

### _Was this patch authored or co-authored using generative AI tooling?_
No.

### CODE REVIEW

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/AccessType.scala [hunk 1]
```diff
@@ -69,7 +69,7 @@ object AccessType extends Enumeration {
               SHOWTABLES |
               SHOWFUNCTIONS |
               DESCFUNCTION => USE
-          case TRUNCATETABLE => UPDATE
+          case TRUNCATETABLE | OPTIMIZETABLE => UPDATE
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/AccessType.scala:72]: IIUC, OPTIMIZETABLE only changes data layout?

Comment (zml1206) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/AccessType.scala:72]: Yes, what type should be used?

Comment (yaooqinn) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/AccessType.scala:72]: It seems that MSCK is approaching.

Comment (zml1206) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/AccessType.scala:72]: Alter access type? Does `VacuumCommand`  also is alter type?

Comment (yaooqinn) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/AccessType.scala:72]: It depends on the impact of Vacuum on data or meta

File: extensions/spark/kyuubi-spark-authz/src/main/resources/table_command_spec.json [hunk 2]
```diff
@@ -2056,6 +2056,30 @@
     "fieldExtractor" : "LogicalPlanQueryExtractor"
   } ],
   "uriDescs" : [ ]
+}, {
+  "classname" : "org.apache.spark.sql.delta.commands.OptimizeTableCommand",
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/resources/table_command_spec.json:2060]: can you provide us the link to this class？

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/AccessType.scala (hunk 1 above)
Comment (zml1206) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/AccessType.scala:72]: `ALTER` is used to modify metadata data of object, `UPDATE` is used to modify physical data of object. `OPTIMIZETABLE` merged the files to generate a new version. Although the data results have not changed, the data files have changed, so is UPDATE more appropriate?

File: extensions/spark/kyuubi-spark-authz/src/main/resources/table_command_spec.json (hunk 2 above)
Comment (zml1206) [reply extensions/spark/kyuubi-spark-authz/src/main/resources/table_command_spec.json:2060]: https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/AccessType.scala (hunk 1 above)
Comment (yaooqinn) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/AccessType.scala:72]: UPDATE is related to record changing

File: extensions/spark/kyuubi-spark-authz/src/main/resources/table_command_spec.json [hunk 3]
```diff
@@ -2056,6 +2056,30 @@
...
+  "tableDescs" : [ {
+    "fieldName" : "tableId",
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/resources/table_command_spec.json:2062]: I couldn't found the field named tableId

Comment (zml1206) [reply extensions/spark/kyuubi-spark-authz/src/main/resources/table_command_spec.json:2062]: In old version. https://github.com/delta-io/delta/blob/v2.2.0/core/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/AccessType.scala (hunk 1 above)
Comment (zml1206) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/AccessType.scala:72]: I understand, change it immediately.

File: extensions/spark/kyuubi-spark-authz/src/main/resources/table_command_spec.json (hunk 3 above)
Comment (yaooqinn) [reply extensions/spark/kyuubi-spark-authz/src/main/resources/table_command_spec.json:2062]: I think we shall target the latest delta release first, then backward compatibilities for elderlies.

Comment (zml1206) [reply extensions/spark/kyuubi-spark-authz/src/main/resources/table_command_spec.json]: Updated

Comment (yaooqinn) [reply extensions/spark/kyuubi-spark-authz/src/main/resources/table_command_spec.json]: Thanks

**Comment by yaooqinn** [comment]
For the test part, I think we shall cover all these cases.

```sql
OPTIMIZE '/path/to/delta/table' -- Optimizes the path-based Delta Lake table

OPTIMIZE delta_table_name;

OPTIMIZE delta.`/path/to/delta/table`;

-- If you have a large amount of data and only want to optimize a subset of it, you can specify an optional partition predicate using `WHERE`:
OPTIMIZE delta_table_name WHERE date >= '2017-01-01'
```

If you don't mind, you can also provide the analyzed logical plans in PR comments for better review experiences, Thank you.

**Comment by zml1206** [comment]
> For the test part, I think we shall cover all these cases.
> 
> ```sql
> OPTIMIZE '/path/to/delta/table' -- Optimizes the path-based Delta Lake table
> 
> OPTIMIZE delta_table_name;
> 
> OPTIMIZE delta.`/path/to/delta/table`;
> 
> -- If you have a large amount of data and only want to optimize a subset of it, you can specify an optional partition predicate using `WHERE`:
> OPTIMIZE delta_table_name WHERE date >= '2017-01-01'
> ```
> 
> If you don't mind, you can also provide the analyzed logical plans in PR comments for better review experiences, Thank you.

Analyzed logical plans added.
I briefly tried it before. Delta Lake table with path is not supported in Authz. I plan to do it separately later. What do you think?

**Comment by yaooqinn** [comment]
Thank you, I'm OK to separated it. Would you mind to attach the plan for `OPTIMIZE '/path/to/delta/table'`?

**Comment by zml1206** [comment]
> Thank you, I'm OK to separated it. Would you mind to attach the plan for `OPTIMIZE '/path/to/delta/table'`?

OPTIMIZE 'tmp/dt1'
== Analyzed Logical Plan ==
path: string, metrics: struct<numFilesAdded:bigint,numFilesRemoved:bigint,filesAdded:struct<min:bigint,max:bigint,avg:double,totalFiles:bigint,totalSize:bigint>,filesRemoved:struct<min:bigint,max:bigint,avg:double,totalFiles:bigint,totalSize:bigint>,partitionsOptimized:bigint,zOrderStats:struct<strategyName:string,inputCubeFiles:struct<num:bigint,size:bigint>,inputOtherFiles:struct<num:bigint,size:bigint>,inputNumCubes:bigint,mergedFiles:struct<num:bigint,size:bigint>,numOutputCubes:bigint,mergedNumCubes:bigint>,numBatches:bigint,totalConsideredFiles:bigint,totalFilesSkipped:bigint,preserveInsertionOrder:boolean,numFilesSkippedToReduceWriteAmplification:bigint,numBytesSkippedToReduceWriteAmplification:bigint,startTimeMs:bigint,endTimeMs:bigint,totalClusterParallelism:bigint,totalScheduledTasks:bigint,autoCompactParallelismStats:struct<maxClusterActiveParallelism:bigint,minClusterActiveParallelism:bigint,maxSessionActiveParallelism:bigint,minSessionActiveParallelism:bigint>,deletionVectorStats:struct<numDeletionVectorsRemoved:bigint,numDeletionVectorRowsRemoved:bigint>,numTableColumns:bigint,numTableColumnsWithStats:bigint>
OptimizeTableCommand DeltaOptimizeContext(false,None,None,false)
+- ResolvedTable org.apache.spark.sql.delta.catalog.DeltaCatalog@608e34bf, delta.`tmp/dt1`, DeltaTableV2(org.apache.spark.sql.SparkSession@3a4af228,tmp/dt1,None,None,None,Map(),org.apache.spark.sql.util.CaseInsensitiveStringMap@1f), [id#818L]

**Comment by zml1206** [comment]
```
OPTIMIZE delta.`/tmp/t1`
before Delta Lake 3.0/Spark 3.5
== Analyzed Logical Plan ==
path: string, metrics: struct<numFilesAdded:bigint,numFilesRemoved:bigint,filesAdded:struct<min:bigint,max:bigint,avg:double,totalFiles:bigint,totalSize:bigint>,filesRemoved:struct<min:bigint,max:bigint,avg:double,totalFiles:bigint,totalSize:bigint>,partitionsOptimized:bigint,zOrderStats:struct<strategyName:string,inputCubeFiles:struct<num:bigint,size:bigint>,inputOtherFiles:struct<num:bigint,size:bigint>,inputNumCubes:bigint,mergedFiles:struct<num:bigint,size:bigint>,numOutputCubes:bigint,mergedNumCubes:bigint>,numBatches:bigint,totalConsideredFiles:bigint,totalFilesSkipped:bigint,preserveInsertionOrder:boolean,numFilesSkippedToReduceWriteAmplification:bigint,numBytesSkippedToReduceWriteAmplification:bigint,startTimeMs:bigint,endTimeMs:bigint,totalClusterParallelism:bigint,totalScheduledTasks:bigint,autoCompactParallelismStats:struct<maxClusterActiveParallelism:bigint,minClusterActiveParallelism:bigint,maxSessionActiveParallelism:bigint,minSessionActiveParallelism:bigint>,deletionVectorStats:struct<numDeletionVectorsRemoved:bigint,numDeletionVectorRowsRemoved:bigint>,numTableColumns:bigint,numTableColumnsWithStats:bigint>
OptimizeTableCommand `delta`.`/tmp/t1`, DeltaOptimizeContext(false,None,None)

Delta Lake 3.0/Spark 3.5
== Analyzed Logical Plan ==
path: string, metrics: struct<numFilesAdded:bigint,numFilesRemoved:bigint,filesAdded:struct<min:bigint,max:bigint,avg:double,totalFiles:bigint,totalSize:bigint>,filesRemoved:struct<min:bigint,max:bigint,avg:double,totalFiles:bigint,totalSize:bigint>,partitionsOptimized:bigint,zOrderStats:struct<strategyName:string,inputCubeFiles:struct<num:bigint,size:bigint>,inputOtherFiles:struct<num:bigint,size:bigint>,inputNumCubes:bigint,mergedFiles:struct<num:bigint,size:bigint>,numOutputCubes:bigint,mergedNumCubes:bigint>,numBatches:bigint,totalConsideredFiles:bigint,totalFilesSkipped:bigint,preserveInsertionOrder:boolean,numFilesSkippedToReduceWriteAmplification:bigint,numBytesSkippedToReduceWriteAmplification:bigint,startTimeMs:bigint,endTimeMs:bigint,totalClusterParallelism:bigint,totalScheduledTasks:bigint,autoCompactParallelismStats:struct<maxClusterActiveParallelism:bigint,minClusterActiveParallelism:bigint,maxSessionActiveParallelism:bigint,minSessionActiveParallelism:bigint>,deletionVectorStats:struct<numDeletionVectorsRemoved:bigint,numDeletionVectorRowsRemoved:bigint>,numTableColumns:bigint,numTableColumnsWithStats:bigint>
OptimizeTableCommand DeltaOptimizeContext(false,None,None)
+- ResolvedTable org.apache.spark.sql.delta.catalog.DeltaCatalog@13553b2e, delta.`/tmp/t1`, DeltaTableV2(org.apache.spark.sql.SparkSession@18f9745a,/tmp/t1,None,None,None,Map()), [id#82, name#83, gender#84, birthDate#85]
```

**Comment by zml1206** [comment]
Only on Delta Lake 3.0/Spark 3.5, the analyzed logical plan of OPTIMIZE delta.`/tmp/t1` and OPTIMIZE '/tmp/t1' are the same.

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/gen/DeltaCommands.scala [hunk 4]
```diff
@@ -53,8 +54,16 @@ object DeltaCommands extends CommandSpecs[TableCommandSpec] {
     TableCommandSpec(cmd, Seq(tableDesc), queryDescs = Seq(queryDesc))
   }
 
+  val OptimizeTableCommand = {
+    val cmd = "org.apache.spark.sql.delta.commands.OptimizeTableCommand"
+    val childDesc = TableDesc("child", classOf[ResolvedTableTableExtractor])
+    val tableDesc = TableDesc("tableId", classOf[TableIdentifierOptionTableExtractor])
+    TableCommandSpec(cmd, Seq(childDesc, tableDesc), OPTIMIZETABLE)
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/gen/DeltaCommands.scala:61]: Let’s use MSCK here?

**Review by yaooqinn — APPROVED** [review]
LGTM, only one minor comment

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/gen/DeltaCommands.scala (hunk 4 above)
Comment (zml1206) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/gen/DeltaCommands.scala:61]: Do you mean to remove the new OperationType of OPTIMIZETABLE and use MSCK directly? Is it better to make a difference? After all, they are two kinds of operations.

Comment (yaooqinn) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/gen/DeltaCommands.scala:61]: Let's keep the enum as a subset of https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/HiveOperationType.java

Comment (zml1206) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/gen/DeltaCommands.scala:61]: How about ALTERTABLE_COMPACT?

Comment (yaooqinn) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/gen/DeltaCommands.scala:61]: Nice choice！

**Review by yaooqinn — APPROVED** [review]

**Comment by zml1206** [comment]
There is a problem with the merge

**Comment by zml1206** [comment]
> There is a problem with the merge

Fixed.

**Comment by yaooqinn** [comment]
Thank you. merged to master

**Comment by zml1206** [comment]
Thanks, @yaooqinn

**Pull request closed without merging**